### PR TITLE
K8SPS-97: Add `delete-backup` finalizer

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPS-97](https://jira.percona.com/browse/K8SPS-97) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPS-97

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/demand-backup.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/auto-config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/gr-init-deploy.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/scaling.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/monitoring.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/config.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/users.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/2490d4a/limits.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/2490d4a9837d0b51d045e5bad429f57cbf8bf438
image: `perconalab/percona-server-mysql-operator:PR-158-2490d4a`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/init-deploy.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/auto-config.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/semi-sync.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/scaling.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/config.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/service-per-pod.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/users.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/5e2c254/limits.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/5e2c2548f20740dae790f3f131bb99ed0edfed5f
image: `perconalab/percona-server-mysql-operator:PR-158-5e2c254`

File: deploy/backup-s3.yaml [hunk 1]
```diff
@@ -2,6 +2,8 @@ apiVersion: v1
 kind: Secret
 metadata:
   name: cluster1-s3-credentials
+  finalizers:
```
Comment (tplavcic) [inline deploy/backup-s3.yaml:5]: We should:
1. add a check in a test that this finalizer is set
2. add a check that when the backup object is deleted from the kubernetes that the backup is actually deleted from the storage (I think we should do it for at least one storage type if not for all)

**Review by tplavcic — CHANGES_REQUESTED** [review]

File: api/v1alpha1/perconaservermysqlbackup_types.go [hunk 2]
```diff
@@ -87,3 +90,31 @@ func (cr *PerconaServerMySQLBackup) Hash() string {
 
 	return hash
 }
+
+type SidecarBackupConfig struct {
```
Comment (inelpandzic) [inline api/v1alpha1/perconaservermysqlbackup_types.go:94]: Why do we now have this SidecarBackupConfig? Why sidecar backup, don't quite understand.

**Review by inelpandzic — COMMENTED** [review]
After backup I have these in my bucket:
- cluster1-2022-09-13-10:13:06-full.md5
- cluster1-2022-09-13-10:13:06-full/

After backup delete I have this:
- cluster1-2022-09-13-10:13:06-full.md5

Should .md5 files be deleted as well?

Also on my initial backup deletion I got these two errors:
```
2022-09-13T10:11:10.006Z        ERROR   checkFinalizers failed to run finalizer delete-backup   {"controller": "perconaservermysqlbackup", "controllerGroup": "ps.percona.com", "controllerKind": "PerconaServerMySQLBackup", "perconaServerMySQLBackup": {"name":"backup1","namespace":"default"}, "namespace": "default", "name": "backup1", "reconcileID": "0438011b-145a-4671-ad67-9b43a28d68b3", "error": "delete backup failed: backup failed\n", "errorVerbose": "delete backup failed: backup failed\n\ngithub.com/percona/percona-server-mysql-operator/controllers.(*PerconaServerMySQLBackupReconciler).deleteBackup\n\t/go/src/github.com/percona/percona-server-mysql-operator/controllers/perconaservermysqlbackup_controller.go:525\ngithub.com/percona/percona-server-mysql-operator/controllers.(*PerconaServerMySQLBackupReconciler).checkFinalizers\n\t/go/src/github.com/percona/percona-server-mysql-operator/controllers/perconaservermysqlbackup_controller.go:353\ngithub.com/percona/percona-server-mysql-operator/controllers.(*PerconaServerMySQLBackupReconciler).Reconcile\n\t/go/src/github.com/percona/percona-server-mysql-operator/controllers/perconaservermysqlbackup_controller.go:119\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1594"}
github.com/percona/percona-server-mysql-operator/controllers.(*PerconaServerMySQLBackupReconciler).Reconcile
        /go/src/github.com/percona/percona-server-mysql-operator/controllers/perconaservermysqlbackup_controller.go:119
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234
2022-09-13T10:11:10.012Z        ERROR   checkFinalizers failed to update finalizers for backup  {"controller": "perconaservermysqlbackup", "controllerGroup": "ps.percona.com", "controllerKind": "PerconaServerMySQLBackup", "perconaServerMySQLBackup": {"name":"backup1","namespace":"default"}, "namespace": "default", "name": "backup1", "reconcileID": "0438011b-145a-4671-ad67-9b43a28d68b3", "backup": "backup1", "error": "Operation cannot be fulfilled on perconaservermysqlbackups.ps.percona.com \"backup1\": StorageError: invalid object, Code: 4, Key: /registry/ps.percona.com/perconaservermysqlbackups/default/backup1, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 64594f69-b1dc-4ba1-8b66-234e451ac0cf, UID in object meta: "}
github.com/percona/percona-server-mysql-operator/controllers.(*PerconaServerMySQLBackupReconciler).Reconcile
        /go/src/github.com/percona/percona-server-mysql-operator/controllers/perconaservermysqlbackup_controller.go:119
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234
```

File: api/v1alpha1/perconaservermysqlbackup_types.go (hunk 2 above)
Comment (pooknull) [reply api/v1alpha1/perconaservermysqlbackup_types.go:94]: We had a struct called `BackupConf` in the `cmd/sidecar.go` file. I just moved it to another package to use it in other places. I renamed it like this because we read from this struct only in the sidecar file. If you think that's not an appropriate name, we can rename it

**Comment by pooknull** [comment]
> After backup I have these in my bucket:
> 
>     * cluster1-2022-09-13-10:13:06-full.md5
> 
>     * cluster1-2022-09-13-10:13:06-full/
> 
> 
> After backup delete I have this:
> 
>     * cluster1-2022-09-13-10:13:06-full.md5
> 
> 
> Should .md5 files be deleted as well?

`md5` files should be deleted, but problem is on the `xbcloud` side

**Comment by inelpandzic** [comment]
> > After backup I have these in my bucket:
> > ```
> > * cluster1-2022-09-13-10:13:06-full.md5
> > 
> > * cluster1-2022-09-13-10:13:06-full/
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > After backup delete I have this:
> > ```
> > * cluster1-2022-09-13-10:13:06-full.md5
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should .md5 files be deleted as well?
> 
> `md5` files should be deleted, but problem is on the `xbcloud` side [PXB-1854](https://jira.percona.com/browse/PXB-1854)

Aha, got it.

File: api/v1alpha1/perconaservermysqlbackup_types.go (hunk 2 above)
Comment (inelpandzic) [reply api/v1alpha1/perconaservermysqlbackup_types.go:94]: What do you think about just naming it like `BackupConf`, because it is just a backup conf and it is being used by the sidecar. Or maybe I'm missing something.

File: controllers/perconaservermysqlbackup_controller.go [hunk 3]
```diff
@@ -315,3 +332,197 @@ func (r *PerconaServerMySQLBackupReconciler) getBackupSource(ctx context.Context
 
 	return src, nil
 }
+
+const (
+	finalizerDeleteBackup = "delete-backup"
```
Comment (inelpandzic) [inline controllers/perconaservermysqlbackup_controller.go:337]: You can remove this const and just use the string directly since it is used only in one place.

File: controllers/perconaservermysqlbackup_controller.go [hunk 4]
```diff
@@ -315,3 +332,197 @@ func (r *PerconaServerMySQLBackupReconciler) getBackupSource(ctx context.Context
...
+	}
+	return nil
+}
+
+func (r *PerconaServerMySQLBackupReconciler) sidecarBackupConfig(ctx context.Context, cr *apiv1alpha1.PerconaServerMySQLBackup) (*apiv1alpha1.SidecarBackupConfig, error) {
+	conf := &apiv1alpha1.SidecarBackupConfig{
+		Destination: cr.Status.Destination,
+		VerifyTLS:   cr.Status.VerifyTLS,
+	}
+	nn := types.NamespacedName{
```
Comment (inelpandzic) [inline controllers/perconaservermysqlbackup_controller.go:379]: It would be better to name it just 'n' or something else because generally in Go repeated letters like this represent a collection. Like `tt := []*Thing`
But it is minor, in this case maybe it is alright. Your call.

File: pkg/xtrabackup/xtrabackup.go [hunk 5]
```diff
@@ -330,6 +432,73 @@ func RestoreJob(
 	}
 }
 
+func DeleteJob(cr *apiv1alpha1.PerconaServerMySQLBackup, conf *apiv1alpha1.SidecarBackupConfig) *batchv1.Job {
```
Comment (inelpandzic) [inline pkg/xtrabackup/xtrabackup.go:435]: What do you think about naming this like `GetDeleteJob` or something like that (maybe `getDeletionJon`), because I initially thought that you want to delete a job with this method :)

File: api/v1alpha1/perconaservermysqlbackup_types.go (hunk 2 above)
Comment (inelpandzic) [reply api/v1alpha1/perconaservermysqlbackup_types.go:94]: I'm looking at this a bit more and if I understand it correctly this config is used only internally, not something related to users. If that is the case maybe it should not be where public API is defined. I think that is the reason why it was initially defined in the sidecar.
If you have to use it, wdyt that you put it in `xtrabackup` package. I feel it is a good fit there.

**Comment by inelpandzic** [comment]
@pooknull did you maybe managed to reproduce the errors I posted earlier?

File: controllers/perconaservermysqlbackup_controller.go (hunk 3 above)
Comment (pooknull) [reply controllers/perconaservermysqlbackup_controller.go]: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/4528a32e0c26a104b9842b349e6439428d2c7bb7

File: pkg/xtrabackup/xtrabackup.go (hunk 5 above)
Comment (pooknull) [reply pkg/xtrabackup/xtrabackup.go]: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/4528a32e0c26a104b9842b349e6439428d2c7bb7

**Comment by pooknull** [comment]
@inelpandzic I reproduced the error. Here is the fix https://github.com/percona/percona-server-mysql-operator/pull/158/commits/2f0ec1c09eec8b85189b2f7cb394e0bf72b8a712

File: api/v1alpha1/perconaservermysqlbackup_types.go (hunk 2 above)
Comment (pooknull) [reply api/v1alpha1/perconaservermysqlbackup_types.go]: When I placed it to `api` package, I thought that it should not be called as `BackupConf` there, because somebody could be confused with `BackupSpec`. But if we move it to the `xtrabackup` package, we can save the old name. I agree that it would be a more appropriate place for this struct. https://github.com/percona/percona-server-mysql-operator/pull/158/commits/4528a32e0c26a104b9842b349e6439428d2c7bb7

File: controllers/perconaservermysqlbackup_controller.go (hunk 4 above)
Comment (pooknull) [reply controllers/perconaservermysqlbackup_controller.go:380]: I'll keep this naming because we use it across the code

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/auto-config.log) |
| demand-backup | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/service-per-pod.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/config.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/limits.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/562c994/users.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/562c994bf21e70846fcfd75f588f3f4cb0d218f6
image: `perconalab/percona-server-mysql-operator:PR-158-562c994`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/demand-backup.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/auto-config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/scaling.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/limits.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/users.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/config.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/fc05cf465363a3e15c51d8268ececca5b5cf581a
image: `perconalab/percona-server-mysql-operator:PR-158-fc05cf4`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/config.log) |
| demand-backup | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/limits.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/scaling.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/fc05cf4/users.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/fc05cf465363a3e15c51d8268ececca5b5cf581a
image: `perconalab/percona-server-mysql-operator:PR-158-fc05cf4`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/auto-config.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/monitoring.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/config.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/limits.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/scaling.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/05fbd18/users.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/05fbd1800e1d96878c9d8841842fe1320a450443
image: `perconalab/percona-server-mysql-operator:PR-158-05fbd18`

File: deploy/backup-s3.yaml (hunk 1 above)
Comment (hors) [inline deploy/backup-s3.yaml:5]: it is the secret why did you add it to the secret

File: deploy/backup.yaml [hunk 6]
```diff
@@ -2,6 +2,8 @@ apiVersion: ps.percona.com/v1alpha1
 kind: PerconaServerMySQLBackup
 metadata:
   name: backup1
+  finalizers:
```
Comment (hors) [inline deploy/backup.yaml:5]: please comment it by default

Comment (pooknull) [reply deploy/backup.yaml]: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/c5333d99b8176fd5a69c8fdb6b927ae0f585f5b3

File: deploy/backup-s3.yaml (hunk 1 above)
Comment (pooknull) [reply deploy/backup-s3.yaml]: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/c5333d99b8176fd5a69c8fdb6b927ae0f585f5b3

**Review by hors — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/auto-config.log) |
| demand-backup | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/service-per-pod.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/config.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/limits.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/scaling.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/c5333d9/users.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/c5333d99b8176fd5a69c8fdb6b927ae0f585f5b3
image: `perconalab/percona-server-mysql-operator:PR-158-c5333d9`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/demand-backup.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/auto-config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/service-per-pod.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/scaling.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/config.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/limits.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-158/8894c00/users.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/158/commits/8894c00da2acfc5a55eef9ab510908e8601b60db
image: `perconalab/percona-server-mysql-operator:PR-158-8894c00`

**Review by hors — APPROVED** [review]

**Review by inelpandzic — APPROVED** [review]

File: controllers/perconaservermysqlbackup_controller.go [hunk 7]
```diff
@@ -315,3 +328,200 @@ func (r *PerconaServerMySQLBackupReconciler) getBackupSource(ctx context.Context
...
+	}
+	sidecarURL := url.URL{
+		Host:   src + ":6033",
+		Scheme: "http",
+		Path:   "/backup/" + cr.Name,
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, sidecarURL.String(), bytes.NewReader(data))
+	if err != nil {
+		return false, errors.Wrap(err, "create http request")
+	}
```
Comment (egegunes) [inline controllers/perconaservermysqlbackup_controller.go:513]: I don't understand why we create a job to delete backup and also send DELETE request to sidecar. They seem to run the same command. What am I missing here?

**Review by egegunes — CHANGES_REQUESTED** [review]
please fix golangci-lint issues, they're critical

File: controllers/perconaservermysqlbackup_controller.go (hunk 7 above)
Comment (pooknull) [reply controllers/perconaservermysqlbackup_controller.go:513]: @egegunes ff a cluster is deleted, there are no available pods to run `xbcloud`, that's why we create a job. Otherwise, we don't need an additional pod and we send `DELETE` to the available pod

Comment (egegunes) [reply controllers/perconaservermysqlbackup_controller.go:513]: I see, thanks for explanation

**Review by egegunes — APPROVED** [review]

**Review by tplavcic — APPROVED** [review]
